Named colors picker

dprokop opened the pull request:
This branch is based on Storybook branch for dev purposes. I'm gonna cherrypick later on before merging. BTW you can run storybook on this branch to play with the components already ;)


This PR grown way larger than initially planned, so here's a summary of major changes:
- Old swatches has been removed from color picker and *named*, limited number of swatches has been added. 
- ColorPicker API (`ColorPickerProps`) and `colorPickerFactory` has been introduced for consistent color picker components
- Pickers has been completely rewritten do React
  - removed `spectrum.js` vendor dependency in favor of `react-color` that's used to build new `SpectrumPalette`
  - popover behavior is now delivered by `popper.js` instead of `drop.js` (unfortunately it's still used by popover_srv.js)
- Named colors(defined in `namedColorsPalette.ts` util, together with helper functions) are now supported. This is an *opt-in* functionality, available via `enableNamedColors` prop from ColorPicker API. Currently it's used in graph legend color pickers only. Opt-in is caused by backward compatibility requirement.
- Barebone `Themable` interface was introduced together with `ConfigProvider` component (delivers boot data via Context API) and `ThemeProvider` (delivers `GraphanaTheme.[Light|Dark]`) type to consumers. This concept is to be revisited when we start Emotion implementation. All react components rendered via `react2AngularDirective` are now wrapped in `ConfigProvider`.
- Deprecation warning for `onColorSelect` prop was added to color pickers


~~Idea is to have composable bits and pieces of color picker that are color-name/hex agnostic. That's why in `@grafana/ui/utils/colorPalette.ts` there are some helper functions that allow color name or definition retrieval based on hex value. `NamedColorPicker` accepts `onChange` prop that will accept entire color definition object. Thanks to that we will be able to retrieve color that's defined for selected theme from Angular components without the need to know what's the current theme on a React component level (i'm not 100% sure if this is a good approach, need to think about it a bit).~~

TODO: 
- [x] Validate concept by using `NamedColorPicker` in e.g. series color picker
- [x]  Figure out approach to theming
- [x] Extract interface for `ColorPicker` component
- [x] Update Angular usage of React color pickers (one API now: `color`, `onChange`)
- [x] Agree on palette and color naming and apply those (@Ijin08)
- [ ] Get rid of inline styles
- [x] Migrate pickers to Popper
- [x] Extend PopperController API to enable hide delay config
- [x] Verify API compatibility with old pickers (for usage in plugins)
- [x] Remove spectrum.js vendor dependency

dprokop committed:
WIP Basics of named color picker

hugohaggmark commented on packages/grafana-ui/src/components/ColorPicker/ColorPicker.story.tsx:
```diff
@@ -1,7 +1,6 @@
 import { storiesOf } from '@storybook/react';
-import { ColorPicker } from '@grafana/ui';
-import { withInfo } from '@storybook/addon-info';
+import { ColorPickerPopover } from './ColorPickerPopover';
```
Maybe rename the file to ColorPickerPopover.story.tsx?

dprokop replied:
Will do!

hugohaggmark commented on packages/grafana-ui/src/components/ColorPicker/ColorPicker.tsx:
```diff
@@ -2,13 +2,15 @@ import React, { Component } from 'react';
    **/
+  name?: Color;
   color: string;
   onChange: (c: string) => void;
```
Maybe rename this to color?

hugohaggmark commented on packages/grafana-ui/src/components/ColorPicker/ColorPicker.tsx:
```diff
@@ -2,13 +2,15 @@ import React, { Component } from 'react';
    **/
+  name?: Color;
   color: string;
   onChange: (c: string) => void;
```
Maybe rename c to color?

hugohaggmark replied:
Should be line 13 :)

dprokop replied:
Yes, that's right

hugohaggmark commented on packages/grafana-ui/src/components/ColorPicker/NamedColorsPicker.tsx:
```diff
@@ -0,0 +1,121 @@
+  ColorsPalete.forEach((colors, hue) => {
+    swatches.push(
+      <>
+        <ColorsGroup selectedColor={selectedColor} colors={colors} onColorSelect={onChange} />
```
A minor question, we don't need to pass a key prop here because we're using a Fragment? (I'm new to the Fragment parts)

dprokop replied:
Nope, just because this is a WIP ;) I simply ignored the warning for now. Good catch!

hugohaggmark reviewed (approved):
I think the PR looks great, I couldn't build the main app because of a missing module 'chalk' but the storybook part looks awesome. Would be great to add Knobs to Storybook as well I think.

Never mind the build error worked second time I tried.

dprokop committed:
Move tooltip themes to Tooltip component making Popper/PopperController theme agnostic

dprokop committed:
Updates to Popper to be positions correctly within window

dprokop committed:
Migrating color pickers to Popper from drop.js pt1

dprokop committed:
Added config provider to be able to access config easily from   react components

dprokop committed:
Render series color picker with correct theme

dprokop committed:
Lint fix

dprokop committed:
Enabled knobs for storybook and implemented some stories

dprokop committed:
Get rid of unused renderContent prop on PopperController

dprokop committed:
Unified color picker API, allowed for color specified for theme selection, updated code to changes in PopperController API

dprokop committed:
Stories updates

dprokop committed:
Fix TS errors

dprokop committed:
Rendering arrows for color picker, applying color changes to time series

dprokop committed:
Updated stories

dprokop committed:
Fix lint

dprokop committed:
Refactor color picker to remove code duplicartion (introduced colorPickerFactory). Allow popver position update on content change

dprokop committed:
Update grafana/ui exports

dprokop committed:
Story updates

dprokop committed:
Make named colors usable in angular code pt 1

dprokop committed:
Updated table tests to new behavior for colors (values are always rendered as hex/rgb)

dprokop committed:
Enable new color picker in Gauge and Thresholds editor, use ThemeProvider instead of ContextSrv

dprokop committed:
Stories - fix import

torkelo commented:
think there is an issue with the scrollbar "track-horizontal" overlaying it making the hit area blow the colored line icon not work, (very small hit area), also think it makes the popover flicker a bit as you hover over it.

The small click area is a problem in the current implementation as well but really annoying and we should try to fix that. 

Regarding migrating old code, that is great, But since this component is used by external plugins we need to make  sure we only enable the "store named colors" feature on demand (so a new prop that turns on this feature). So current plugins will get the hex/rgba color and only with "namedColorsEnabled=true" property does the onChange carry a named color

dprokop commented:
@torkelo I have already talked with @daniellee about externals plugins, and we are going to investigate pickers usage out there.

dprokop commented:
> think there is an issue with the scrollbar "track-horizontal" overlaying it making the hit area blow the colored line icon not work, (very small hit area), also think it makes the popover flicker a bit as you hover over it.

@torkel are you referring to series color picker?

torkelo commented:
> > think there is an issue with the scrollbar "track-horizontal" overlaying it making the hit area blow the colored line icon not work, (very small hit area), also think it makes the popover flicker a bit as you hover over it.
> 
> @torkel are you referring to series color picker?

yes

dprokop committed:
Make popover hide delay configurable to enable better UX

dprokop committed:
CustomScrollbar - expose underlying's react-custom-scrollbars API to allow scroll tracks config.

Updated config for Graph legend not to display horizontal scroll track which caused legend being not accessible in ceertain situations (e.g. when native OS scroll overlay disabled)

dprokop commented:
@torkelo, the issue with Legend not being accessible is now fixed with dd8f7aa7a9c7f7240b92c347006ff8df7e91bdd5.

dprokop committed:
Storybook - add actions addon

dprokop committed:
Make named colors optional in color picker, enable named colors in graph legend series picker

dprokop committed:
Stories update for color picker

dprokop commented:
@torkelo added `enableNamedColors` prop to ColorPicker API and enabled named colors in graph legend for now. Gonna add some tests for that.

jschill commented:
Nice!

1. Would it be an idea to have the white "selected" circle outside of the color'd circle instead? White doesnt play well with the light colors, but it will always look good against the black bg.
2. Is there gonna be a close button? (and Esc?)
3. Nothing important, but maybe it would be nice to see the new color when you hover the circle, like a preview, so we could reduce the amount of clicks. Might be overkill, but i think it would be a smoother ride.

dprokop committed:
Implemented new spectrum palette

dprokop committed:
Reduce padding in color picker popover

dprokop committed:
Added deprecation warning to old color picker API props. Moved named color support handling to color popovers

dprokop committed:
Make default color picker close on trigger mouse leave

dprokop committed:
Implement pointer component for spectrum palette sliders

dprokop committed:
Update styles of selected named color swatch

dprokop commented:
@jschill 
> Would it be an idea to have the white "selected" circle outside of the color'd circle instead? White doesnt play well with the light colors, but it will always look good against the black bg.

Well, it works pretty well with dark theme. But with light it's barely visible:
![image](https://user-images.githubusercontent.com/2376619/51611569-d5078d80-1f1f-11e9-8195-0707027f4a88.png)
![image](https://user-images.githubusercontent.com/2376619/51611581-ddf85f00-1f1f-11e9-9d9b-6df47e1415d5.png)

> Is there gonna be a close button? (and Esc?)

We have not planned for that in this iteration. 

> Nothing important, but maybe it would be nice to see the new color when you hover the circle, like a preview, so we could reduce the amount of clicks. Might be overkill, but i think it would be a smoother ride.

I really like this idea! Can you put it into backlog please? We can iterate over it later on.

dprokop committed:
Make small swatches react to theme changes

dprokop committed:
Rename colorsPalette util to namedColorsPalette

dprokop committed:
Fix hide timeout for color picker

dprokop committed:
Remove spectrum.js vendor dependency from grafana/ui

Ijin08 committed:
replaced palette colors with current palette adjusted for dark and light theme

Ijin08 committed:
Merge branch 'named-colors' of github.com:grafana/grafana into named-colors

dprokop committed:
Removed sass import of spectrum.scss

torkelo commented on packages/grafana-ui/src/components/ColorPicker/ColorPickerPopover.tsx:
```diff
@@ -1,111 +1,86 @@
-import tinycolor from 'tinycolor2';
-import { ColorPalette } from './ColorPalette';
-import { SpectrumPicker } from './SpectrumPicker';
+import NamedColorsPicker from './NamedColorsPalette';
```
looks like there was a rename here that did not get updated here. Does this need a default export?

dprokop replied:
Nope, will fix

torkelo commented:
looks really great. 

when used from the series override it over.

![2019-01-23_18-56-59_screenshot](https://user-images.githubusercontent.com/10999/51626829-b82f8200-1f40-11e9-9493-02bedec9ef6a.png)

I kind of like moving the y-axis to it's own tab:
![2019-01-23_18-44-51_screenshot](https://user-images.githubusercontent.com/10999/51626855-c41b4400-1f40-11e9-80fe-79d78d83c91b.png)

thoughts?

torkelo commented:
The selected color state looks broken in both singlestat & Gauge panel:  
![image](https://user-images.githubusercontent.com/10999/51627042-3d1a9b80-1f41-11e9-9a21-0964599de99c.png)

dprokop commented:
> The selected color state looks broken in both singlestat & Gauge panel:

That's because of spectrum.scss being removed now. Sorry!

dprokop commented:
> when used from the series override it overflows

It's not migrated yet. First thing in the morning! That's exactly the same issue that's present in pie char plugin -> it's not the picker that is used, but picker's popover (in drop.js context which has max width to 20rem set...)

dprokop committed:
Fixed issue with color name retrieval not being aware of current theme

dprokop committed:
Restore missing styles

dprokop committed:
Update imports of NamedColorsPalette

dprokop committed:
Update story for NamedColorsPalette

dprokop committed:
Make series overrides color picker display correctly

dprokop commented:
@torkelo 
> when used from the series override it overflows

This is fixed now, but I have created an issue to fix it for good both in pie-chart an series override https://github.com/grafana/grafana/issues/15038 , https://github.com/grafana/piechart-panel/pull/167

dprokop committed:
Enable custom picers on color color picker popovers (for y-axis support in legend picker)

jschill commented:
> > Would it be an idea to have the white "selected" circle outside of the color'd circle instead? White doesnt play well with the light colors, but it will always look good against the black bg.
> 
> Well, it works pretty well with dark theme. But with light it's barely visible:
Right. I was thinking the circle could be black on the light theme.

Ijin08 committed:
minor styling changes to gaps, font-size and width

dprokop committed:
Move Switch component to grafana-ui

dprokop committed:
Use Switch to control y-axis in series color picker

dprokop committed:
Stories cleanup

dprokop committed:
Fix light theme issues with named colors disabled

dprokop committed:
Implemented tests for ColorPickerPopover and NamedColorsPalette

dprokop committed:
Rename deprecation warning helper

dprokop committed:
Do not update color picker popover position on tab change

Position update caused tabs to move up and down. This led to confusing user experience, as tabs position was dependent on tab's content.

torkelo commented:
looking really good! For the pickers I am missing a cursor: pointer that was there before

dprokop committed:
Add cursor pointer to color swatches

dprokop commented:
> For the pickers I am missing a cursor: pointer that was there before


fixed

Ijin08 committed:
updated the color palette

torkelo reviewed (commented):
Wound one bug and that is the Graph thresholds (setting at the bottom), it has a custom color option that is not working 100%. Seems like the change is not taking effect. The popover is shown but the selecting is not propagated to the graph. 

Think it's because line color is undefined for new thresholds and is causing an exception when undefined is passed to getColorFromHexRgbOrName.

torkelo commented on packages/grafana-ui/src/utils/namedColorsPalette.ts:
```diff
@@ -0,0 +1,163 @@
+ColorsPalette.set('red', reds);
+ColorsPalette.set('blue', blues);
+ColorsPalette.set('orange', oranges);
+ColorsPalette.set('purple', purples);
```
We did something similar to unit formats, that was previously also built in global scope, moved that to be done on demand instead: https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/utils/valueFormats/valueFormats.ts#L134

dprokop committed:
Fix thresholds default colors not being applied

dprokop committed:
Initialize named colors palete lazily

dprokop committed:
Merge branch 'named-colors' of github.com:grafana/grafana into named-colors
